compiler/domain.py

```diff
@@ -1,84 +1,162 @@
-    PDF = "pdf"   # type: ignore
-    DVI = "dvi"   # type: ignore
-    PS = "ps"   # type: ignore
+class Status(Enum):
+    """Represents the status of a requested compilation."""
 
-    CURRENT = "current"   # type: ignore
-    IN_PROGRESS = "in_progress"   # type: ignore
-    FAILED = "failed"   # type: ignore
+    COMPLETED = "completed"
+    IN_PROGRESS = "in_progress"
+    FAILED = "failed"
+
+
+class Reason(Enum):
+    """Specific reasons for a (usually failure) outcome."""
+
+    AUTHORIZATION = "auth_error"
+    MISSING = "missing_source"
+    SOURCE_TYPE = "invalid_source_type"
+    CORRUPTED = "corrupted_source"
+    STORAGE = "storage"
+    CANCELLED = "cancelled"
+    ERROR = "compilation_errors"
+    NETWORK = "network_error"
+    NONE = None
+
+
+class Task(NamedTuple):
+    """Represents the state of a compilation product in the store."""
 
     # Here are the actual slots/fields.
-    source_id: str
+    status: Status
+    """
+    The status of the compilation.
 
-    format: str
+    If :attr:`Status.COMPLETED`, the current file corresponding to the format
+    of this compilation status is the product of this compilation.
+    """
+
+    source_id: Optional[str] = None
+
+    output_format: Optional[Format] = None
     """
     The target format of the compilation.
 
     One of :attr:`PDF`, :attr:`DVI`, or :attr:`PS`.
     """
 
-    source_checksum: str
-    """Checksum of the source tarball from the file management service."""
+    checksum: Optional[str] = None
+    """
+    Checksum of the source tarball from the file management service.
 
-    task_id: str
+    This is likely to be a checksum of some kind, but may be something else.
+    """
+
+    task_id: Optional[str] = None
     """If a task exists for this compilation, the unique task ID."""
 
-    status: str
-    """
-    The status of the compilation.
+    reason: Reason = Reason.NONE
+    """An explanation of the current status. E.g. why did it fail."""
 
-    One of :attr:`CURRENT`, :attr:`IN_PROGRESS`, or :attr:`FAILED`.
+    description: str = ""
+    """A description of the outcome."""
 
-    If :attr:`CURRENT`, the current file corresponding to the format of this
-    compilation status is the product of this compilation.
-    """
+    size_bytes: int = 0
+    """Size of the product."""
+
+    owner: Optional[str] = None
+    """The owner of this resource."""
 
     @property
     def ext(self) -> str:
         """Filename extension for the compilation product."""
-        return self.format
+        if self.output_format is None:
+            raise TypeError('Output format `None` has no extension')
+        return self.output_format.ext
+
+    @property
+    def content_type(self) -> str:
+        """Mime type for the output format of this compilation."""
+        if self.output_format is None:
+            raise TypeError('Output format `None` has no content type')
+        return self.output_format.content_type
 
     def to_dict(self) -> dict:
```

Comment:
In this case couldn't you just return `self._asdict()`?

Reply:
I did it this way so that I could convert the enum values to ``str``, e.g.

```python
    'output_format':
                self.output_format.value if self.output_format else None,
    ...
    'status': self.status.value if self.status else None,
    'reason': self.reason.value if self.reason else None,
```